miner/worker_builder.go

```diff
@@ -0,0 +1,450 @@
+	var mu sync.Mutex
+	for i, bundle := range bundles {
+		if simmed, ok := simCache.GetSimulatedBundle(bundle.Hash()); ok {
+			mu.Lock()
+			simResult[bundle.Hash()] = simmed
+			mu.Unlock()
+			continue
+		}
+
+		wg.Add(1)
+		go func(idx int, bundle *types.Bundle, state *state.StateDB) {
+			defer wg.Done()
+			gasPool := prepareGasPool()
+			simmed, err := w.simulateBundle(env, bundle, state, gasPool, 0, true, true)
+			if err != nil {
+				log.Trace("Error computing gas for a simulateBundle", "error", err)
+				return
+			}
+
+			mu.Lock()
+			defer mu.Unlock()
+			simResult[bundle.Hash()] = simmed
+		}(i, bundle, env.state.Copy())
+	}
+
+	wg.Wait()
+
+	simulatedBundles := make([]*types.SimulatedBundle, 0)
+
+	for _, bundle := range simResult {
+		if bundle == nil {
+			continue
+		}
+
+		simulatedBundles = append(simulatedBundles, bundle)
+	}
+
+	simCache.UpdateSimulatedBundles(simResult, bundles)
+
+	return simulatedBundles, nil
+}
+
+// mergeBundles merges the given simulateBundle into the given environment.
+// It returns the merged simulateBundle and the number of transactions that were merged.
+func (w *worker) mergeBundles(
+	env *environment,
+	bundles []*types.SimulatedBundle,
+) (types.Transactions, *types.SimulatedBundle, error) {
+	currentState := env.state.Copy()
+	gasPool := prepareGasPool()
+	env.UnRevertible = mapset.NewSet[common.Hash]()
+
+	includedTxs := types.Transactions{}
+	mergedBundle := types.SimulatedBundle{
+		BundleGasFees:  new(big.Int),
+		BundleGasUsed:  0,
+		BundleGasPrice: new(big.Int),
+	}
+
+	for _, bundle := range bundles {
+		prevState := currentState.Copy()
+		prevGasPool := new(core.GasPool).AddGas(gasPool.Gas())
+
+		// the floor gas price is 99/100 what was simulated at the top of the block
+		floorGasPrice := new(big.Int).Mul(bundle.BundleGasPrice, big.NewInt(99))
+		floorGasPrice = floorGasPrice.Div(floorGasPrice, big.NewInt(100))
+
+		simulatedBundle, err := w.simulateBundle(env, bundle.OriginalBundle, currentState, gasPool, len(includedTxs), true, false)
+
+		if err != nil && errors.Is(err, core.ErrGasLimitReached) {
+			log.Error("failed to merge bundle, interrupt merge process", "err", err)
+			break
+		}
+		if err != nil || simulatedBundle.BundleGasPrice.Cmp(floorGasPrice) <= 0 {
+			currentState = prevState
+			gasPool = prevGasPool
+
+			log.Error("failed to merge bundle", "floorGasPrice", floorGasPrice, "err", err)
+			continue
+		}
+
+		log.Info("included bundle",
+			"gasUsed", simulatedBundle.BundleGasUsed,
+			"gasPrice", simulatedBundle.BundleGasPrice,
+			"txcount", len(simulatedBundle.OriginalBundle.Txs))
+
+		includedTxs = append(includedTxs, bundle.OriginalBundle.Txs...)
+
+		mergedBundle.BundleGasFees.Add(mergedBundle.BundleGasFees, simulatedBundle.BundleGasFees)
+		mergedBundle.BundleGasUsed += simulatedBundle.BundleGasUsed
+
+		for _, tx := range includedTxs {
+			if !containsHash(bundle.OriginalBundle.RevertingTxHashes, tx.Hash()) {
+				env.UnRevertible.Add(tx.Hash())
```

Comment:
if this happen, the bundle should be dropped, but I didn't find this logic.

Reply:
Use this logic In commitBundleTransaction, drop bundle: `receipt.status == ReceiptStatusFailed && unRevertible`.